pkg/sql/colexec/table_function/plugin.go

```diff
@@ -0,0 +1,329 @@
+			}
+
+			cfgmap[string(key)] = string(value)
+			return nil
+		})
+	}
+	//logutil.Infof("ARGS %v", args)
+
+	// datalink
+	dlVec := tf.ctr.argVecs[3]
+	if dlVec.IsNull(uint64(nthRow)) {
+		u.batch.SetRowCount(0)
+		return nil
+	}
+	src := dlVec.GetStringAt(nthRow)
+
+	var bytes []byte
+	switch dlVec.GetType().Oid {
+	case types.T_datalink:
+		dl, err := datalink.NewDatalink(src, proc)
+		if err != nil {
+			return err
+		}
+		bytes, err = dl.GetPlainText(proc)
+		if err != nil {
+			return err
+		}
+	case types.T_varchar, types.T_text, types.T_char:
+		bytes = []byte(src)
+	default:
+		return moerr.NewInternalError(proc.Ctx, "plugin_exec input type not supported")
+	}
+
+	wurl, err := url.Parse(wasmurl)
+	if err != nil {
+		return err
+	}
+
+	if u.plugin == nil {
+		var manifest extism.Manifest
+		if wurl.Scheme == "https" || wurl.Scheme == "http" {
+
+			manifest = extism.Manifest{
+				Wasm: []extism.Wasm{
+					extism.WasmUrl{
+						Url: wasmurl,
+					},
+				},
+				Config:       cfgmap,
+				AllowedHosts: u.allowed_hosts,
+			}
+		} else {
+			// treat as datalink
+			wasmdl, err := datalink.NewDatalink(wasmurl, proc)
+			if err != nil {
+				return err
+			}
+			image, err := wasmdl.GetBytes(proc)
+			if err != nil {
+				return err
+			}
+
+			manifest = extism.Manifest{
+				Wasm: []extism.Wasm{
+					extism.WasmData{
+						Data: image,
+					},
+				},
+				Config:       cfgmap,
+				AllowedHosts: u.allowed_hosts,
+			}
+		}
+
+		config := extism.PluginConfig{
+			EnableWasi:                true,
+			EnableHttpResponseHeaders: true,
+		}
+		u.plugin, err = extism.NewPlugin(proc.Ctx, manifest, config, []extism.HostFunction{})
+		if err != nil {
+			return err
+		}
+	}
+
+	exit, out, err := u.plugin.Call(funcname, bytes)
+	if err != nil {
+		return moerr.NewInternalError(proc.Ctx, fmt.Sprintf("plugin exit with error %d. err %v", exit, err))
+	}
+
+	nitem := 0
+	var jserr error
+	jsonparser.ArrayEach(out, func(value []byte, dataType jsonparser.ValueType, offset int, err error) {
+
+		if err != nil {
+			jserr = errors.Join(jserr, err)
+			return
+		}
+
+		bj := bytejson.ByteJson{}
+		isnull := false
+		switch dataType {
+		case jsonparser.String:
+			jsdata, err := json.Marshal(string(value))
+			if err != nil {
+				jserr = errors.Join(jserr, err)
+				return
+			}
+			err = bj.UnmarshalJSON(jsdata)
+			if err != nil {
+				jserr = errors.Join(jserr, err)
+				return
+			}
+		case jsonparser.Number, jsonparser.Boolean, jsonparser.Object, jsonparser.Array:
+			err = bj.UnmarshalJSON(value)
+			if err != nil {
+				jserr = errors.Join(jserr, err)
+				return
+			}
+		case jsonparser.Null:
+			isnull = true
+		default:
+			jserr = errors.Join(jserr, moerr.NewInternalErrorNoCtx("unknown json type"))
+			return
+		}
+		vector.AppendByteJson(u.batch.Vecs[0], bj, isnull, proc.Mp())
+		nitem += 1
+	})
+
+	if jserr != nil {
+		return jserr
+	}
+
+	u.batch.SetRowCount(nitem)
```

Comment:
basically, if here nitem is 0, we should also be able to process.

Reply:
Also does the table function has a "end of input" call?    Basically, if I want to use table function to implement my own group by, I will process each input row, but for each input row, I don't output anything.   I just keep on updating the internal state.    When all input has been processed, call "end of input", in this call I will output the groups in my internal state.

Reply:
This probably means you have to update/fix the table_function.go framework.   We are missing this "end of input" state/flag and need to check the flag and call.

Reply:
Can you give example in SQL?  CROSS APPLY and JOIN don't work like this.